### PR TITLE
Feature/dynamic content

### PR DESCRIPTION
**Pull request by manuel-pross**
## Description
This PR adds the rendering logic of the homepage (hero and body).
This includes:
- typing of strapi responses
- components for the different strapi fields
- styling of the different strapi fields displayed in the frontend

What dies PR does NOT include:

- Two column layout of some homepage content (we should implement sth. for that in the strapi page builder). Refer: https://miro.com/app/board/uXjVNKsIEgU=/
- Mobile design implementation

## What type of PR is this? (check all applicable)
- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [x] 🎨 Style
- [x] 🧑‍💻 Code Refactor
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore
- [ ] ⏩ Revert

## Related Issues and Pull Requests
**Fixes:** #(issue/ pr number) <br/>
**Closes:** #(issue/ pr number) <br/>
**Depends on:** #(issue/ pr number)

### Desktop
https://github.com/vim/website_next_generation/assets/48802396/fa9f43a5-1200-4a0a-9733-42db15914d35

## [optional] Design Documents or Screenshots

https://miro.com/app/board/uXjVNKsIEgU=/?moveToWidget=3458764578884341560&cot=14

### CODE REVIEW

File: web/src/app/page.tsx
Comment (jomunker) [inline web/src/app/page.tsx:11]: We could also think about using [SWR](https://www.npmjs.com/package/swr) for data fetching

File: web/src/components/Strapi/Common/Card.tsx
Comment (jomunker) [inline web/src/components/Strapi/Common/Card.tsx:15]: this will result in `className=false` in the dom if `hasShadow` is false. Use a ternary operator for this please.

Either:
```suggestion
		<div className={`${hasShadow ? "flex flex-col rounded pt-7 pb-4 px-10 bg-gray-4 drop-shadow" : ""}`}>
```
Or (but please try this before committing the suggestion here):
```suggestion
		<div className={hasShadow ? "flex flex-col rounded pt-7 pb-4 px-10 bg-gray-4 drop-shadow" : undefined}>
```

File: web/src/components/Strapi/Sections/Content.tsx
Comment (jomunker) [inline web/src/components/Strapi/Sections/Content.tsx:15]: Please use a nicer fallback text here.

File: web/src/components/Strapi/Sections/Content.tsx
Comment (jomunker) [inline web/src/components/Strapi/Sections/Content.tsx:20]: Why not also use a switch statement here?
you can combine cases like this => 
```tsx
switch(entry.type) {
  case "Card":
  case "Plain":
  case "Accordion":
    return // ...
  // ...
}
```

File: web/src/components/Strapi/Sections/Content.tsx
Comment (jomunker) [inline web/src/components/Strapi/Sections/Content.tsx:54]: is the fragment here needed?

**Review by jomunker — COMMENTED** [review]
Looks good so far 🙌🏻  I'll wait for the screen recording with my approval.

File: cms/src/api/home/content-types/home/schema.json
Comment (manuel-pross) [reply cms/src/api/home/content-types/home/schema.json:3]: It confuses me too. This is autogenerated from strapi. I removed it and it does not seem to have any side effects

File: web/src/app/api/[homepage]/route.ts
Comment (manuel-pross) [reply web/src/app/api/[homepage]/route.ts:20]: We have a component called 'NewsSection' in our strapi backend. It defines a Card with the amount of news (presumably the latest news) which you want to display at your page.

The news objects are not in the newscount. For every newscount element in the page content we have to fetch the news accordingly.

Maybe we can enhance that in the future with caching the already fetched news.

File: cms/src/api/home/controllers/home.ts
Comment (manuel-pross) [reply cms/src/api/home/controllers/home.ts:12]: I copied this code from another branch without questioning it in the first place.
I wondered during testing the strapi response for the homepage why always all news entry from the collection type news are sent too.

This line is the cause of that. So thanks for pointing it out :)

My personal opinion is that the response for questioning the homepage data should not contain all news per default. For every news section defined in the content there should be a separate request to fetch the requested amount of news. In the future we should improve that with proper caching

File: web/src/app/page.tsx
Comment (manuel-pross) [reply web/src/app/page.tsx:11]: I had to do this because we had some pipeline issues regarding our build step. The API was not available there and I had to implement a route handler to fetch it without issues.

File: web/src/components/Strapi/Sections/Content.tsx
Comment (manuel-pross) [reply web/src/components/Strapi/Sections/Content.tsx:15]: Done

File: web/src/components/Strapi/Common/Card.tsx
Comment (manuel-pross) [reply web/src/components/Strapi/Common/Card.tsx:15]: The second suggestion results in class=undefined
The first one in an empty class.
Could not observe any visible difference in the rendered result.
Implemented the first one.

File: cms/src/api/home/content-types/home/schema.json
Comment (jomunker) [reply cms/src/api/home/content-types/home/schema.json:3]: Alright then commit it 👍🏻

File: cms/src/api/home/controllers/home.ts
Comment (jomunker) [reply cms/src/api/home/controllers/home.ts:12]: Yeah makes sense to use the same imo. Makes it easier to understand. Yeah we should have pagination for news. Also, for the homepage we only need the last 5 or so news articles so I'd restrict it to that.

File: web/src/app/api/[homepage]/route.ts
Comment (jomunker) [reply web/src/app/api/[homepage]/route.ts]: Tbh I don't get the code here. Seems really confusing and not self explanatory. Can you simplify it?

File: web/src/app/page.tsx
Comment (jomunker) [reply web/src/app/page.tsx:11]: Yes SWR is also for client components but offers a very nice UX. Maybe take a look. It's also recommended by Next.js

File: cms/src/api/home/content-types/home/schema.json
Comment (jomunker) [inline cms/src/api/home/content-types/home/schema.json:5]: I'd not remove the `collectionName` but rather change it to home. Also I'd rename the `pluralName` to home

File: cms/src/api/home/content-types/home/schema.json
Comment (manuel-pross) [reply cms/src/api/home/content-types/home/schema.json]: Done

File: web/src/app/api/[homepage]/route.ts
Comment (jomunker) [inline web/src/app/api/[homepage]/route.ts:22]: Small input. You could use a flatMap instead of map. Then you don't need the filter function :) Just do if ... else return []

File: web/src/app/api/[homepage]/route.ts
Comment (jomunker) [reply web/src/app/api/[homepage]/route.ts]: Maybe also rename some variables

**Comment by jomunker** [comment]
videoframe_4437
 The right side should be also rounded at the bottom and the rounded window should not scroll away as it does right now. But this is probably not a concern for this PR.

File: cms/src/api/home/controllers/home.ts
Comment (manuel-pross) [reply cms/src/api/home/controllers/home.ts]: I do not know if 5 is just an example. The page maintainer could decide to increase the amount of news displayed.
For that case the getNews-Funktion has a default value of 5 for the count parameter

Comment (manuel-pross) [reply cms/src/api/home/controllers/home.ts]: We can limit the maximal possible value for the count in our strapi news section component if you referring to that

File: cms/src/api/home/content-types/home/schema.json
Comment (manuel-pross) [reply cms/src/api/home/content-types/home/schema.json]: Results in an error:

2024-05-02 15:17:07 Error: The plural name "home" should be unique

File: web/src/app/api/[homepage]/route.ts
Comment (manuel-pross) [reply web/src/app/api/[homepage]/route.ts]: Did not thought about that. Thanks!

Done ;)

File: web/src/app/api/[homepage]/route.ts
Comment (manuel-pross) [reply web/src/app/api/[homepage]/route.ts]: Did a refactoring. Pls. let me know if this is more understandable now.

The "newsCount" attribute is used for type narrowing. This field exists only in the NewsSection-Type which is part of the union type GenericContentEntry.

So this code ensures that only news sections are used for fetching of the news.

File: web/src/app/page.tsx
Comment (manuel-pross) [reply web/src/app/page.tsx]: Implemented it in a very simple way

File: cms/src/api/home/controllers/home.ts
Comment (jomunker) [reply cms/src/api/home/controllers/home.ts]: Yeah 5 was just an example. I mean that on the homepage just 5 or whatever should be delivered in total because more are not needed.

File: cms/src/api/home/content-types/home/schema.json
Comment (jomunker) [reply cms/src/api/home/content-types/home/schema.json:6]: rip okay^^

**Comment by jomunker** [comment]
pls resolve conflicts

File: web/package.json
Comment (jomunker) [inline web/package.json:23]: 👍🏻

File: web/src/app/page.tsx
Comment (jomunker) [inline web/src/app/page.tsx:10]: we should add a ROUTES enum probably

**Review by jomunker — APPROVED** [review]

**Review by jomunker — APPROVED** [review]

**Pull request merged**
